test/parallel/test-https-end-rst.js

```diff
@@ -0,0 +1,31 @@
+'use strict';
+const common = require('../common');
+
+const http = require('http');
+const net = require('net');
+
+const data = 'hello';
+const server = http.createServer({}, common.mustCallAtLeast((req, res) => {
+  res.setHeader('content-length', data.length);
+  res.end(data);
+}, 2));
+
+server.listen(0, function() {
+  const options = {
+    host: '127.0.0.1',
+    port: server.address().port,
+    allowHalfOpen: true
+  };
+  const socket = net.connect(options, common.mustCall(() => {
+    socket.on('ready', common.mustCall(() => {
+      socket.write('GET /\n\n');
+      socket.once('data', common.mustCall(() => {
+        socket.write('GET /\n\n');
+        setTimeout(common.mustCall(() => {
+          socket.destroy();
+          server.close();
+        }), common.platformTimeout(10));
```

Comment:
Is the timeout needed? Could we use some event on the other side?

Reply:
You won't get anything on the second request - the server has already closed its end
The previous version will send a `RST`
The new one will silently swallow your request

Reply:
@mcollina The more I look at this, the more I am starting to question this behavior
Because now it is correct from the socket standpoint, but not from the protocol
Normally, an HTTP/1.1 connection with `Connection: close` should have a mandatory simultaneous close
In this case, force-closing and sending back a `RST` if the other end continues writing is the correct behavior

We got the TLS right, but not HTTP/1.1

Reply:
But for this to work, we have to correctly identify the exact protocol version
Then enforce a mandatory simultaneous close in HTTP/1.1 and TLS/1.0
And drain the input side on HTTP/1.0 and TLS/1.1+

Reply:
@addaleax @lpinca @jasnell 
What do you think?

Reply:
@ronag

Reply:
In fact, HTTP, be it 1.0 or 1.1 does not specify any closing mechanism
TLS 1.0 specifies a mandatory simultaneous close after exchanging two `close-notify` messages
As this was one of the most disrespected aspects of the specification (most implementation did force close the socket), they made it optional in TLS 1.1 where one-sided closing is officially supported - the other must ignore everything sent after the `close-notify`

So in theory:
* The old behavior was correct for TLS 1.0
* The new one is correct for TLS 1.1
* HTTP is undefined

In practice:
* The new one does not get `RST`s
* The old one did

Reply:
How about force-closing in HTTP and drain-closing in TLS? Simple to implement, no `RST` and almost 100% correct - most implementations do not close properly in TLS 1.0 anyway

Reply:
I made a small program to test HTTP server behavior: <https://gist.github.com/mmomtchev/918b70c93a56bc83fc7ada2066c77123>

All servers I tested except Google's own server have the new behavior (after this changeset)
Google's own HTTP server has the old behavior

Reply:
It is quite random in fact. So, yes, HTTP's behavior in this case is undefined

Reply:
In fact, my test is not reliable at all because most servers do not use `Connection: close`, never mind

Reply:
I see some of issues in your gist:

1. There is no guarantee that the whole response will be read in a single `'data'` event.
1. The `try...catch` is statement is not needed. No error is thrown synchronously.
1. You can send the `Connection: close` header from the client to tell the server to close the socket.

Reply:
For the test, I think it is possible to remove the timeout by listening to the `'finish'` event of response socket. When that event is emitted write again from the client socket.

Reply:
@lpinca I have no `response` object, I am doing manual HTTP with `net`?

I fixed the gist and I ran with a Node without the `readStop` kludge in `stream_base_commons.js` - it was still not very reliable detecting the `RST` until I added a third `write` - ie. if you call `socket.write()` to send some data and that data gets a `RST` in return you won't get an error until you `read` - which you won't since the `Readable` has ended - or call another `write`. Since this is very edge behavior anyway, I won't call it a bug.

The test is still very random - usually naked HTTP servers (ie Apache) will read after `FIN` while the big load-balancers of the top sites won't

Reply:
> I have no `response` object, I am doing manual HTTP with `net`?

You have it on the server. Your server is an HTTP server, you client is a plain `net.Socket`. Basically wait for the `'finish'` event to be emitted on the `response`, then wait for the `'finish'` event on the server socket and when it is emitted trigger another write from the client socket.

Reply:
The reason for this is that Node ignores `SIGPIPE` on Linux - normally if you call `write` (the kernel call) in non-blocking mode, it will just schedule your write and return. This is what libuv does too. Then later, if this `write` gets a `RST` in return - the only way to notify you would be a `SIGPIPE` - if you don't do any other calls on this descriptor and you have ignored `SIGPIPE` you won't get an error. But once again, this is a very edge behavior, I don't think it is worth it to try to fix it.

Anyway, this confirms that HTTP's behavior, when it comes to closing the connection, is undefined and both implementations exist in the wild

Reply:
With the extra `write` the gist should work even with a stock Node